docs/folialint.1
.TH folialint 1 "2020 sep 02" "version 2.0 "
.
.SH NAME
folialint \(hy check validity of FoLiA documents
.
.SH SYNOPSIS
folialint [options] FILE(S)
.
.SH DESCRIPTION
.
.B folialint
will check the validity of FoLiA files on a surface level:

\(hy Is the tag structure according to the definition?

\(hy Are all required elements and attributes present?

\(hy Are the attributes and values valid?

\(hy Are all annotations correctly declared?

\(hy Is all text consistent?
.
.SH OPTIONS
.
.B --output
file
.RS
write output to 'file'. (default is stdout).
Assumes only 1 inputfile.
.RE
.
.B --nooutput
.RS
suppress al output except errors
.RE
.
.B --nochecktext
.RS
Disable checking text consistency. (for older FoLiA versions this wasn't forced to be so. Which is
.B wrong
)
.RE
.
.B --fixtext
.RS
Attempt to fix text problems like wrong offset's. Might be dangerous, and the
default is to DON'T do it.
.RE
.
.B -a
or
.B --autodeclare
.RS
Attempt to automaticly fix missing annotations. (default: false)
.RE
.
.B --strip
.RS
strip all information from the file that makes comparison hard.
Notably date/time information.
.
This also enable '--canonical' output, meaning that otherwise unsorted tags
are output in the same order every time.
.
.RE
.
.B --permissive
.RS
Accept documents with missing namespace declarations.
.RE
.
.B --warn
.RS
add some extra (warning) checks:

\(hy About unused set declarations

\(hy About possible incompatabilities between the current library version and
the document version.
.RE
.
.B --canonical
or
.B --KANON
.RS
output annotation declarations and FoLiA nodes in a fixed (canonical) order.
This might make tree comparisons easier.
.RE
.
.B -d
or
.B --debug level
.RS
run with a debug level of
.B level
.RE
.
.B -V
or
.B --version
.RS
Show VERSION
.RE
.
.B -h
or
.B --help
.RS
Show some help
.RE
.
.SH BUGS
possible
.
.SH AUTHORS
Ko van der Sloot: lamasoftware@science.ru.nl


NEWS
2.5 2020-09-02
[Maarten van Gompel]
* Adapted to FoLiA v2.3
* Support parsing of the new explicit form
[Ko vd Sloot]
* folialint: updated usage() and man page
* minor refactoring

2.4 2020-04-15
[Ko van der Sloot]
* comment in Doxygen format added
* bumped the library version to 14
* fix for https://github.com/proycon/folia/issues/82
* fix for https://github.com/proycon/folia/issues/42
* fixed problem with using new tag names on pre 1.6 documents
* better checks in folia_engine on text inconsistencies and such
  (https://github.com/LanguageMachines/libfolia/issues/43)
* confidence output is more consistent now
* removed the folia_builder (was not used)
* code refactorings and cleanup, removing unused functions

2.3.2 2020-01-13
[Ko van der Sloot]
Bug fix release
* fix for https://github.com/LanguageMachines/foliautils/issues/37
* fix for https://github.com/LanguageMachines/foliautils/issues/38
* fixes in Correction handling.
* fixed a Multi-Threading problem with the static reverse_old map

2.3.1 2019-10-21
[Maarten van Gompel]
* Bug fix release for gcc 9.1
  It stumbles upon some inline functions
[Ko van der Sloot]
* replaced call to unsafe 'tmpnam()' by 'TiCC::tempname()'

2.3 2019-08-29
[Ko vd Sloot]
new features:
* autodeclare mode introduced (as in FoLiApy)
* folialint by default doesn't autodeclare. use -a or --autodeclare to use it
* Better detecting of declaration errors in general
* the select function now also enables the possibility to search recursively
  upto the first matching sibling.
* bumped library version to 13

other changes:
* some exceptions are changed.
* less exceptions are thrown. An empty result is returned instead.
* folialint now accept bote -d and --debug
* real fix for issue 70
* small bug fixes and refactorings
* accept empty Comment and Description nodes

2.2.1 2019-07-22
[Ko vd Sloot]
Bug fix release:
* There were some problems handling NO setname vs. EMPTY setname, during
  incremental parsing in folia::Engine. This was sorted out now:
  https://github.com/proycon/folia/issues/74
  This related to some ucto and frog issues too:
  https://github.com/LanguageMachines/ucto/issues/70
  https://github.com/LanguageMachines/frog/issues/72

2.2 2019-07-15
[Ko vd Sloot]
Bug fix release.
* Folia::Engine choked on some complex FoLiA. Solved by refactoring and in fact
  simplifying some code. (Frog issue #77 revealed this)
* added flush() on document output to streams. (frog issue #72)
* improved output in debugging mode

2.1 2019-06-19
[Ko vd Sloot]
Bug fixes and enhancements:
* provenance:
 - added 'generate_id' attribute with 'auto()' and 'next()' values
 - some code improvements
* bugs:
 - When using the FoLiA-engine, we have to save the ORIGINAL annotationdefaults,
   and use these when parsing.

2.0 2019-05-22
[Ko vd Sloot]
Major release.
* Supports the new FoLiA 2.0 features:
  - provenance support
  - more stricter checking on annotation declarations
  - added the new TextMarkupReference class
  - supports Hidden Words.
  - All structure elements can have the 'space' attribute
  - support for groupannotations
  - many more.

* API and ABI breaches:
  - libray version bumped to version 10
  - many functions are renamed
  - the text() functions have an ENUM parameter now to select for STRICT,
    RETAIN or HIDDEN

* bug fixes
  - support for xlink: improved
  - there as a rare mixup between <text> nodes and <text-annotation> nodes in
    the folia::Engine
  - all <wref> nodes get a 't' attribute now on serializing.
  - reading Extrenal Folia could getint an endless loop

* code refactoring and cleanup

1.16 2019-05-15
[Ko van der Sloot]
stabilizing release for folia1.5. Next release wil support the new FoLiA 2.0

Changes:
* renamed folia::Processor to folia::Engine
* extended and improved Engine code a lot
* avoid spurious newline on Document output
* Will read and ignore some FoLiA 2.0 additions
* numerous small additions and fixes
* make sure that the XmlParser uses the HUGE model everywhere

1.15 2018-11-29
[Ko van der Sloot]
* added (still experimental) code for a FoLiA Builder, Processor and
  TextProcessor class.
  Use with care. The API may change unannounced!
* a foliadiff script (using folialint) is installed now
* several refactorings, to make the code more clear.
* the 'ref' attribute was not serialized for TextContent
* several smaller small bug fixes
* the .so version is bumped to 9 because of a lot of API/ABI changes

1.14 never released

1.13 2018-05-16
[Ko van der Sloot]
* disabled WordRefrence test. It was incomplete, and hard to do
* use icu:: namespace

[Maarten van Gompel]
* added codemeta.json
* fix spelling errors in error messages

1.12 2018-02-19
* configuration cleanup. MacOSX is better supported now.
* folialint now supports --fixtext (handle with care!)
* library version bumped to 8.0, due to changes in the API
* regenerated FoLiA properties (to FoLiA version 1.5.1)
* several small bug fixes

1.11 2017-12-04
Bug fix release:
* handling of <comment> tags within <t> nodes
* better handling of <wref> tags. Forbid forward references

1.10.1 2017-11-06
Minor fix
* bumped the .so version to 7.0

1.10 2017-10-17
Major Release, implementing FoLiA spec 1.5
* added text checking for all 1.5 documents and up
* added offset and ref checking for Text in all 1.5 documents and up
* 'empty' text inside TextContent, PhonContent and Textmarkup is significamt
* better version checking
* text checking can be dis/enabled using FOLIA_TEXT_CHECK environment variable
* added submetadata mechanism
* implemented aliasses for annotation setnames
* added an xmlstring() serializer for Document
* bug fixes:
  - in LineBreak serializing
  - XmlComment is textless.
  - miscelaneous small fixes

1.9  2017-08-30
Bug fix release
* accept ICU 50 too (was 52) to make CentOS happy
* XmlComment INSIDE <t> lead to crashes. fixed.
* code changes in code that is only executed for documents in folia 1.5 format
  (that shouldn't exist in the wild)

1.8  2017-00-16

src/folialint.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of libfolia

  libfolia is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  libfolia is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/ticcutils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/

#include <iostream>
#include <fstream>
#include <string>
#include <map>
#include "ticcutils/CommandLine.h"
#include "libfolia/folia.h"

using namespace std;

void usage(){
  cerr << "usage: folialint [options] <foliafiles>" << endl;
  cerr << "options are" << endl;
  cerr << "\t-h, --help\t\t This help" << endl;
  cerr << "\t-V, --version\t\t Show versions" << endl;
  cerr << "\t--strip\t\t\t strip variable items from the FoLiA. Like all dates." << endl;
  cerr << "\t\t\t\t This is usefull to generate FoLiA that can be diffed." << endl;
  cerr << "\t--output='file'\t\t name an outputfile. (default is stdout)" << endl;
  cerr << "\t--nooutput\t\t Suppress output. Only warnings/errors are displayed." << endl;
  cerr << "\t--nochecktext\t\t DO NOT check if text is consistent inside structure tags." << endl;
  cerr << "\t\t\t\t Default is to do so." << endl;
  cerr << "\t--fixtext.\t\t Try to fixup text errors like wrong offsets." << endl;
  cerr << "\t\t\t\t Default is to DON'T DO THAT." << endl;
  cerr << "\t-a, --autodeclare\t Attempt to automaticly fix missing annotations." << endl;
  cerr << "\t\t\t\t (default: false)" << endl;
  cerr << "\t--permissive.\t\t Allow some dubious constructs." << endl;
  cerr << "\t--warn\t\t\t add some extra warnings about library versions and unused" << endl;
  cerr << "\t\t\t\t annotation declarations" << endl;
  cerr << "\t--canonical\t\t output in a predefined order. Makes comparisons easier" << endl;
  cerr << "\t--KANON\t\t\t same as --canonical" << endl;
  cerr << "\t-d value, --debug=value\t Run more verbose." << endl;
  cerr << "\t--permissive\t\t Accept some unwise constructions." << endl;
}

int main( int argc, char* argv[] ){
  string outputName;
  bool permissive;
  bool warn;
  bool strip ;
  bool nooutput = false;
  bool nochecktext = false;
  bool fixtext = false;
  bool kanon = false;
  bool autodeclare = false;
  string debug;
  vector<string> fileNames;
  string command;
  try {
    TiCC::CL_Options Opts( "hVd:a",
			   "nochecktext,debug:,permissive,strip,output:,nooutput,help,fixtext,warn,version,canonical,KANON,autodeclare");
    Opts.init(argc, argv );
    if ( Opts.extract( 'h' )
	 || Opts.extract( "help" ) ){
      usage();
      return EXIT_SUCCESS;
    }
    if ( Opts.extract( 'V' )
	 || Opts.extract( "version" ) ){
      cout << "folialint version 0.5" << endl;
      cout << "based on [" << folia::VersionName() << "]" << endl;
      return EXIT_SUCCESS;
    }
    command = Opts.toString();
    if ( !command.empty() ){
      command = "folialint " + command;
    }
    else {
      command = "folialint";
    }
    permissive = Opts.extract("permissive");
    warn = Opts.extract("warn");
    nooutput = Opts.extract("nooutput");
    fixtext = Opts.extract("fixtext");
    kanon = Opts.extract("canonical") || Opts.extract("KANON");
    if ( Opts.extract("nochecktext") ){
      nochecktext = true;
    }
    strip = Opts.extract("strip");
    if ( strip && permissive ){
      cerr << "conflicting options: 'permissive' and 'strip'" << endl;
      return EXIT_FAILURE;
    }
    Opts.extract( "debug", debug ) || Opts.extract( 'd', debug );
    Opts.extract( "output", outputName );
    autodeclare = Opts.extract( "autodeclare" ) || Opts.extract( 'a' );

    if ( !Opts.empty() ){
      cerr << "unsupported option(s): " << Opts.toString() << endl;
      return EXIT_FAILURE;
    }
    fileNames = Opts.getMassOpts();
    if ( fileNames.size() == 0 ){
      cerr << "missing input file" << endl;
      usage();
      return EXIT_FAILURE;
    }
    if ( fileNames.size() > 1 && !outputName.empty() ){
      cerr << "--outputname not supported for more then 1 inputfile" << endl;
      return EXIT_FAILURE;
    }
    string inputName = fileNames[0];
    if ( inputName == outputName ){
      cerr << "output name cannot be same as input name!" << endl;
      return EXIT_FAILURE;
    }
  }
  catch( exception& e ){
    cerr << "FAIL: " << e.what() << endl;
    exit( EXIT_FAILURE );
  }

  int fail_count = 0;
  string mode;
  if ( permissive ){
    mode += ",permissive";
  }
  if ( strip ){
    mode += ",strip";
  }
  if ( fixtext ){
    mode += ",fixtext";
  }
  if ( autodeclare ){
    mode += ",autodeclare";
  }
  else {
    mode += ",noautodeclare"; // the default
  }
  if ( nochecktext ){
    mode += ",nochecktext";
  }
  else {
    mode += ",checktext";	// the default
  }
  if ( !mode.empty() ){
    mode = ", mode='" + mode + "'";
  }
  if ( !debug.empty() ){
    mode += ", debug='" + debug + "'";
  }
  for ( const auto& inputName : fileNames ){
    try {
      string cmd = "file='" + inputName + "'";
      cmd += mode;
      //      cerr << "running " << cmd << endl;
      folia::Document d( cmd );
      if ( !d.version_below(2,0)
	   && !(kanon||strip)
	   && d.get_processors_by_name( "folialint" ).empty() ){
	folia::KWargs args;
	args["name"] = "folialint";
	args["id"] = "folialint";
	args["generator"] = "yes";
	args["begindatetime"] = "now()";
	args["command"] = command;
	folia::processor *proc = d.add_processor( args );
	proc->get_system_defaults();
	proc->set_metadata( "valid", "yes" );
      }
      if ( !outputName.empty() ){
	d.save( outputName, kanon );
      }
      else if ( !nooutput ){
	cout << d;
      }
      else {
	cerr << "Validated successfully: " << inputName << endl;
      }
      if ( warn ){
	if ( d.compare_to_build_version() ){
	  cerr << "WARNING: the document had version: " << d.version()
	       << " and the library is at version: "
	       <<  folia::folia_version() << endl;
	}
	multimap<folia::AnnotationType, string> und = d.unused_declarations();
	if ( !und.empty() ){
	  cerr << "the following annotationsets are declared but unused: " << endl;
	  for ( const auto& it : und ){
	    cerr << folia::toString( it.first )<< "-annotation, set=" << it.second << endl;
	  }
	}
      }
    }
    catch( exception& e ){
      cerr << inputName << " failed: " << e.what() << endl;
      ++fail_count;
    }
  }
  if ( fail_count > 0 ){
    exit( EXIT_FAILURE );
  }
  exit( EXIT_SUCCESS );
}
